### PR TITLE
Polish grpcio_health_checking package

### PR DESCRIPTION
**Pull request by kpayson64**
-Rename namespace to grpc_health->grpc to match spec
-Proper use of NOT_FOUND status code
-Improve testing
-Add source distribution to artifact build

### CODE REVIEW

File: src/python/grpcio_health_checking/grpc/health/v1/health.py [hunk 1]
```diff
@@ -44,10 +46,7 @@ def __init__(self):
   def Check(self, request, context):
     with self._server_status_lock:
       if request.service not in self._server_status:
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_health_checking/grpc/health/v1/health.py]: Eliminate the double-dipping into the dictionary:
```
status = self._server_status.get(request.service)
if status is None:
  context.set_code(grpc.StatusCode.NOT_FOUND)
  return health_pb2.HealthCheckResponse()
else:
  return health_pb2.HeathCheckResponse(status=status)
```
.

_Technically_ we don't support fail-to-return-a-value-from-a-response-unary-method-implementation-but-it's-all-cool-because-we-set-a-non-OK-code. It happens to behave compatibly today but let's not put it into this code until after we've decided to certify it as a proper feature and added test coverage?

File: src/python/grpcio_tests/tests/health_check/_health_servicer_test.py [hunk 2]
```diff
@@ -27,49 +27,70 @@
...
 
 class HealthServicerTest(unittest.TestCase):
 
   def setUp(self):
-    self.servicer = health.HealthServicer()
-    self.servicer.set('', health_pb2.HealthCheckResponse.SERVING)
-    self.servicer.set('grpc.test.TestServiceServing',
+    servicer = health.HealthServicer()
+    servicer.set('', health_pb2.HealthCheckResponse.SERVING)
+    servicer.set('grpc.test.TestServiceServing',
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_tests/tests/health_check/_health_servicer_test.py]: Arguments to call are misaligned and [I don't care](https://github.com/grpc/grpc/issues/262). :-)

File: src/python/grpcio_tests/tests/health_check/_health_servicer_test.py [hunk 3]
```diff
@@ -27,49 +27,70 @@
...
 
   def test_not_serving_service(self):
     request = health_pb2.HealthCheckRequest(
         service='grpc.test.TestServiceNotServing')
-    resp = self.servicer.Check(request, None)
+    resp = self._stub.Check(request)
     self.assertEqual(resp.status, health_pb2.HealthCheckResponse.NOT_SERVING)
 
+  def test_not_found_service(self):
+    request = request = health_pb2.HealthCheckRequest(
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_tests/tests/health_check/_health_servicer_test.py]: `request = request = <expression>`? Isn't this a parse error in Python?

File: src/python/grpcio_tests/tests/health_check/_health_servicer_test.py [hunk 4]
```diff
@@ -27,49 +27,70 @@
...
+        service='not-found')
+    with self.assertRaises(grpc.RpcError) as context:
+      resp = self._stub.Check(request)
+  
+    print(context.exception.details())
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_tests/tests/health_check/_health_servicer_test.py]: Unit tests should not print.

File: src/python/grpcio_tests/tests/health_check/_health_servicer_test.py [hunk 5]
```diff
@@ -27,49 +27,70 @@
...
+    self.assertEqual(context.exception.code(), grpc.StatusCode.NOT_FOUND)
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_tests/tests/health_check/_health_servicer_test.py]: Python doesn't define an order like the unit test libraries of other languages do but throughout our codebase we've tried to maintain `<expected value>, <value obtained from system under test>` for the arguments to `assertEqual`. Please continue that?

On comment proofreading: yes, this means changing otherwise-untouched lines in this test.

File: src/python/grpcio_tests/tests/health_check/_health_servicer_test.py (hunk 3 above)
Comment (kpayson64) [inline src/python/grpcio_tests/tests/health_check/_health_servicer_test.py]: It works in the interpreter (but yes its a typo :) )

**Comment by kpayson64** [comment]
Addressed comments.

**Comment by nathanielmanistaatgoogle** [comment]
Luhguhtum.

Does this resolve, or make progress on, an issue?

**Comment by kpayson64** [comment]
This is running into some issues with the artifact build, so I'm removing the ready to merge label.

**Comment by kpayson64** [comment]
I've gotten the artifact build working:
https://grpc-testing.appspot.com/view/Artifacts/job/gRPC_build_artifacts/1275/

Re-assigning to @soltanmm for artifact build changes review.

File: tools/distrib/python/grpcio_tools/grpc/tools/command.py [hunk 6]
```diff
@@ -35,6 +35,25 @@
 from grpc.tools import protoc
 
 
+def build_package_protos(package_root):
```
Comment (soltanmm) [inline tools/distrib/python/grpcio_tools/grpc/tools/command.py]: Could we either rename this to `build_proto_modules` or rename `BuildProtoModules` to `BuildPackageProtos` (my preference is on the former, but whatever's fine as long as the two correspond with each other)?

File: tools/run_tests/build_artifact_python.sh [hunk 7]
```diff
@@ -81,5 +81,19 @@ then
   done
 fi
 
+# We need to use the built grpcio-tools/grpcio to compile the health proto
+# Wheels are not supported by setup_requires/dependency_links, so we
```
Comment (soltanmm) [inline tools/run_tests/build_artifact_python.sh]: Do you mean that `setup_requires` doesn't retrieve wheels or that our wheels in our directory aren't retrieved by `setup_requires` and there's no way to configure that to happen?

File: tools/run_tests/build_artifact_python.sh [hunk 8]
```diff
@@ -81,5 +81,19 @@ then
...
+# manually install the dependency
+if [ "$SKIP_PIP_INSTALL" == "" ]
```
Comment (soltanmm) [inline tools/run_tests/build_artifact_python.sh]: I think `$SKIP_PIP_INSTALL` is more an optimization-hint-thing; if we need to install the built `grpcio-tools` for correctness then we probably want to skip `$SKIP_PIP_INSTALL`.

File: tools/distrib/python/grpcio_tools/grpc/tools/command.py [hunk 9]
```diff
@@ -52,19 +71,4 @@ def run(self):
...
-    for proto_file in proto_files:
-      command = [
-          'grpc.tools.protoc',
-          '--proto_path={}'.format(inclusion_root),
-          '--python_out={}'.format(inclusion_root),
-          '--grpc_python_out={}'.format(inclusion_root),
-      ] + [proto_file]
-      if protoc.main(command) != 0:
-        sys.stderr.write('warning: {} failed'.format(command))
+    build_package_protos(self.distribution.package_dir[''])
```
Comment (soltanmm) [inline tools/distrib/python/grpcio_tools/grpc/tools/command.py]: Someday there'll be an option to configure a different package directory, or maybe a way to just walk all package directories without tripping on anything unforeseen. Someday.

File: tools/run_tests/build_artifact_python.sh (hunk 7 above)
Comment (kpayson64) [inline tools/run_tests/build_artifact_python.sh]: Unfortunately, setup_requires doesn't retrieve wheels period (https://github.com/pypa/setuptools/issues/377)

File: tools/run_tests/build_artifact_python.sh (hunk 8 above)
Comment (kpayson64) [inline tools/run_tests/build_artifact_python.sh]: I wanted to avoid install grpcio/grpcio-tools if we are not in a virtualenv or docker, so I ended up using this variable to enforce that.  I've renamed it to BUILD_HEALTH_CHECKING

**Comment by kpayson64** [comment]
Addressed comments.

Updated build
https://grpc-testing.appspot.com/view/Artifacts/job/gRPC_build_artifacts/1318/

File: tools/run_tests/build_artifact_python.sh (hunk 7 above)
Comment (soltanmm) [inline tools/run_tests/build_artifact_python.sh]: 😶

File: tools/run_tests/build_artifact_python.sh [hunk 10]
```diff
@@ -39,15 +39,6 @@ export PIP=${PIP:-pip}
 export AUDITWHEEL=${AUDITWHEEL:-auditwheel}
 
 
-if [ "$SKIP_PIP_INSTALL" == "" ]
```
Comment (soltanmm) [inline tools/run_tests/build_artifact_python.sh]: Why was this removed? When I said 'optimization' earlier, I meant that it was an optimization to skip this step, not an optimization to install.

Unless we no longer break when we don't have these lines? I guess we'll see in the watched build.

Comment (kpayson64) [inline tools/run_tests/build_artifact_python.sh]: Our setup_requires should take care of these.  SKIP_PIP_INSTALL was disabled for both mac and linux before this PR (and is not used for windows)

**Comment by kpayson64** [comment]
Build passed.

**Comment by soltanmm** [comment]
LGTM
